rxlifecycleexampleapp/src/main/java/co/uk/conjure/rxlifecycle/exampleapp/LoginViewModel.kt

```diff
@@ -0,0 +1,173 @@
+package co.uk.conjure.rxlifecycle.exampleapp
+
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewmodel.initializer
+import androidx.lifecycle.viewmodel.viewModelFactory
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
+import io.reactivex.rxjava3.core.Observable
+import io.reactivex.rxjava3.core.Observer
+import io.reactivex.rxjava3.kotlin.addTo
+import io.reactivex.rxjava3.subjects.BehaviorSubject
+import io.reactivex.rxjava3.subjects.PublishSubject
+import uk.co.conjure.rxlifecycle.RxViewModel
+
+/**
+ * The ViewModel interface for the Login.
+ * It only offers Observables and Observers for the UI.
+ *
+ */
+interface LoginViewModel {
+    // input
+    val email: Observer<String>
+    val password: Observer<String>
+    val loginClick: Observer<Unit>
+
+    // output
+    val isLoginButtonEnabled: Observable<Boolean>
+    val isLoading: Observable<Boolean>
+    val error: Observable<LoginError>
+}
+
+/**
+ * Navigation event when the Login completes.
+ * It is separate from the [LoginViewModel] because it's not the Views job to perform the navigation.
+ */
+interface LoginNavigationModel {
+    val onLoginComplete: Observable<Unit>
+}
+
+enum class LoginError {
+    NONE,
+    INVALID_CREDENTIALS,
+    GENERIC_ERROR
+}
+
+/**
+ * Actual implementation of the [LoginViewModel].
+ * It also implements [LoginNavigationModel] which will be observed by the Activity to navigate to
+ * the next screen.
+ */
+class LoginViewModelImpl(private val api: LoginApi) : RxViewModel(), LoginViewModel,
+    LoginNavigationModel {
+
+    override val email: PublishSubject<String> = PublishSubject.create()
+    override val password: PublishSubject<String> = PublishSubject.create()
+    override val loginClick: PublishSubject<Unit> = PublishSubject.create()
+
+    private val state: BehaviorSubject<LoginState> =
+        BehaviorSubject.createDefault(LoginState.Idle("", ""))
+
+
+    private val userInput = Observable.merge(
+        email.map { UiEvent.EmailChange(it) },
+        password.map { UiEvent.PasswordChange(it) },
+        loginClick.map { UiEvent.LoginClick }
+    )
+
+    init {
+        userInput.withLatestFrom(state) { event, state ->
+            updateState(state, event)
+        }
+            .distinctUntilChanged()
+            .publish { currentState ->
+                Observable.merge(
+                    currentState,
+                    performLogin(currentState)
+                )
+            }
+            .subscribe { state.onNext(it) }
+            .addTo(keepAlive)
+    }
+
+
+    override val isLoginButtonEnabled: Observable<Boolean> = state.map {
+        (it is LoginState.Idle || it is LoginState.Error) &&
+                it.email.isNotBlank() && it.password.isNotBlank()
+    }
+
+    override val isLoading: Observable<Boolean> =
```

Comment:
Not super necessary but maybe we should show the use of hot here? .. or somewhere in this class.